File: commons-csv-graal/src/main/java/org/apache/commons/csv/ExtendedBufferedReader.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.commons.csv;

import static org.apache.commons.csv.Constants.CR;
import static org.apache.commons.csv.Constants.END_OF_STREAM;
import static org.apache.commons.csv.Constants.LF;
import static org.apache.commons.csv.Constants.UNDEFINED;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.Reader;

import org.graalvm.polyglot.Value;

/**
 * A special buffered reader which supports sophisticated read access.
 * <p>
 * In particular the reader supports a look-ahead option, which allows you to see the next char returned by
 * {@link #read()}. This reader also tracks how many characters have been read with {@link #getPosition()}.
 * </p>
 */
final class ExtendedBufferedReader extends BufferedReader {

    /** glue code attributes */
    private static Value clz = ContextInitializer.getPythonClass("extended_buffered_reader.py", "ExtendedBufferedReader"); 
    private Value obj;

    public Value getPythonObject() {
        return obj;
    }

    /**
     * Created extended buffered reader using default buffer-size
     */
    ExtendedBufferedReader(final Reader reader) {
        super(reader);
        obj = clz.newInstance(reader, true);
    }

    /**
     * Converts a Python character to an integer value.
     * If value is a string, it invokes the "_ord" method in Python.
     * If the given value is not a string, it directly coerces to an integer value.
     *
     * @param v the value to be converted
     * @return the integer representation of the Python character
     */
    private int convertPythonCharToInt(Value v) {
        if (v.isString())
            return (int) v.asString().charAt(0);
        else
            return v.asInt();
    }

    @Override
    public int read() throws IOException {
       return convertPythonCharToInt(obj.invokeMember("read"));
    }

    /**
     * Returns the last character that was read as an integer (0 to 65535). This will be the last character returned by
     * any of the read methods. This will not include a character read using the {@link #lookAhead()} method. If no
     * character has been read then this will return {@link Constants#UNDEFINED}. If the end of the stream was reached
     * on the last read then this will return {@link Constants#END_OF_STREAM}.
     *
     * @return the last character that was read
     */
    int getLastChar() {
        // return lastChar;
        return convertPythonCharToInt(obj.invokeMember("get_last_char"));
    }

    @Override
    public int read(final char[] buf, final int offset, final int length) throws IOException {
        return convertPythonCharToInt(obj.invokeMember("read", buf, offset, length));
    }

    /**
     * Calls {@link BufferedReader#readLine()} which drops the line terminator(s). This method should only be called
     * when processing a comment, otherwise information can be lost.
     * <p>
     * Increments {@link #eolCounter}
     * <p>
     * Sets {@link #lastChar} to {@link Constants#END_OF_STREAM} at EOF, otherwise to LF
     *
     * @return the line that was read, or null if reached EOF.
     */
    @Override
    public String readLine() throws IOException {
        return obj.invokeMember("read_line").asString();
    }

    /**
     * Returns the next character in the current reader without consuming it. So the next call to {@link #read()} will
     * still return this value. Does not affect line number or last character.
     *
     * @return the next character
     *
     * @throws IOException
     *             if there is an error in reading
     */
    int lookAhead() throws IOException {
        return convertPythonCharToInt(obj.invokeMember("look_ahead"));
    }

    /**
     * Returns the current line number
     *
     * @return the current line number
     */
    long getCurrentLineNumber() {
        return obj.invokeMember("get_current_line_number").asLong();
    }

    /**
     * Gets the character position in the reader.
     *
     * @return the current position in the reader (counting characters, not bytes since this is a Reader)
     */
    long getPosition() {
        return obj.invokeMember("get_position").asLong();
    }

    public boolean isClosed() {
        return obj.invokeMember("is_closed").asBoolean();
    }

    /**
     * Closes the stream.
     *
     * @throws IOException
     *             If an I/O error occurs
     */
    @Override
    public void close() throws IOException {
        obj.invokeMember("close");
    }

}

File: commons-csv-python/src/main/python/extended_buffered_reader.py
import io
from types import MethodType
from main.python.constants import Constants
from main.python.java_handler import Types


class ExtendedBufferedReader(io.StringIO):
    def __init__(self, reader, from_java=False):
        self._last_char = Constants.UNDEFINED
        self._eol_counter = 0
        self._position = 0
        self._closed = False
        self.mark = None
        self._write_pos = 0
        self.from_java = from_java
        
        # init super class based on callee environment
        if self.from_java:
            self.reader = reader
            self.buffered_reader = Types.BufferedReader(reader)
        else:
            super().__init__(reader.read())
            __write = reader.write
            def _write(this, s: str) -> int:
                self.seek(self._write_pos)
                self.write(s)
                self._write_pos = self.tell()
                self.seek(self._position)
                return __write(s)
            reader.write = MethodType(_write, reader)
            self.reader = None
            self.buffered_reader = None

    def _readline_python(self, length=-1):
        initial_pos = self.tell()
        value = super().readline(length)

        if not value:
            return None

        offset = 0
        if value.endswith("\r\n"):
            value = value[:-2]
            offset = 2
        if value.endswith(("\r", "\n")):
            value = value[:-1]
            offset = 1
            
        min_break = len(value)
        if "\r" in value and value.index("\r") < min_break:
            min_break = value.index("\r")
            offset = 1
        if "\n" in value and value.index("\n") < min_break:
            min_break = value.index("\n")
            offset = 1
        if "\r\n" in value and value.index("\r\n") <= min_break:
            min_break = value.index("\r\n")
            offset = 2

        s = value[:min_break]
        self.seek(initial_pos + len(s) + offset)
        return s

    def _readline(self, length=-1):
        if self.from_java:
            return self.buffered_reader.readLine()
        else:
            return self._readline_python(length)
    
    def _read(self, length=None):
        if self.from_java:
            char_int = self.buffered_reader.read()
            return chr(char_int) if char_int != -1 else Constants.END_OF_STREAM
        else:
            value = super().read(length)
            return value if value else Constants.END_OF_STREAM

    def read(self, *args):
        if len(args) == 0:
            current = self._read(1)
            if current == Constants.CR or (current == Constants.LF and self._last_char != Constants.CR):
                self._eol_counter += 1
            self._last_char = current
            self._position += 1
            return self._last_char      
        if len(args) == 3:
            buf, offset, length = args
            i = 0
            c = self.read()
            if c == Constants.END_OF_STREAM and length > 0:
                return Constants.END_OF_STREAM
            
            while i < length and c != Constants.END_OF_STREAM:
                buf[offset + i] = c
                i += 1
                if i < length:
                    c = self.read()
            return i

    def get_last_char(self):
        return self._last_char

    def read_extended(self, length=None):
        if length == 0:
            return 0

        buf = super().read(length)

        if buf:
            for ch in buf:
                if ch == Constants.LF:
                    if self._last_char != Constants.CR:
                        self._eol_counter += 1
                elif ch == Constants.CR:
                    self._eol_counter += 1

            self._last_char = buf[-1]

        elif buf == "":
            self._last_char = Constants.END_OF_STREAM

        self._position += len(buf)
        return len(buf)

    def read_line(self):
        line = self._readline()
        if line or line == "":
            self._last_char = Constants.LF
            self._eol_counter += 1
        else:
            self._last_char = Constants.END_OF_STREAM
        return line

    def look_ahead(self):
        if self.from_java:
            self.buffered_reader.mark(1)
        else: 
            self.mark = self.tell()
        
        c = self._read(1)
        self.reset()

        return c
    
    def reset(self):
        if self.from_java:
            self.buffered_reader.reset()
        else:
            if self.mark == None:
                raise IOError("Mark not set")
            self.seek(self.mark)        

    def get_current_line_number(self):
        if (
            self._last_char == Constants.CR
            or self._last_char == Constants.LF
            or self._last_char == Constants.UNDEFINED
            or self._last_char == Constants.END_OF_STREAM
        ):
            return self._eol_counter
        return self._eol_counter + 1

    def get_position(self):
        return self._position

    def is_closed(self):
        return self._closed

    def close(self):
        self._closed = True
        self._last_char = Constants.END_OF_STREAM
        if self.from_java:
            self.buffered_reader.close()
            self.reader.close()
        else:
            super().close()
